Webpack Build Error: Can't Resolve 'fs' and 'stream' in payloadcms/plugin-cloud-storage

I'm trying to use the payloadcms plugin-cloud-storage in my project and I encounter errors when webpack tries to build the project.

The error messages are as follows:
```ERROR in ./node_modules/@payloadcms/plugin-cloud-storage/dist/adapters/azure/handleUpload.js 44:27-40
Module not found: Error: Can't resolve 'fs' in 'C:\dev\tempo\TempoStormWeb\cms-app\node_modules\@payloadcms\plugin-cloud-storage\dist\adapters\azure'
ERROR in ./node_modules/@payloadcms/plugin-cloud-storage/dist/adapters/azure/handleUpload.js 45:15-32
Module not found: Error: Can't resolve 'stream' in 'C:\dev\tempo\TempoStormWeb\cms-app\node_modules\@payloadcms\plugin-cloud-storage\dist\adapters\azure'
```
Here's my package.json:
```
{
  "name": "cms-app",
  "version": "1.0.0",
  "private": "true",
  "resolutions": {
    "sharp": "latest"
  },
  "scripts": {
    "dev": "cross-env PAYLOAD_CONFIG_PATH=src/payload.config.ts NODE_ENV=development nodemon",
    "build:payload": "cross-env PAYLOAD_CONFIG_PATH=src/payload.config.ts payload build",
    "build:server": "tsc",
    "build": "yarn copyfiles && yarn build:payload && yarn build:server",
    "serve": "cross-env PAYLOAD_CONFIG_PATH=dist/payload.config.js NODE_ENV=production node dist/server.js",
    "copyfiles": "copyfiles -u 1 \"src/**/*.{html,css,scss,ttf,woff,woff2,eot,svg,jpg,png}\" dist/",
    "generate:types": "cross-env PAYLOAD_CONFIG_PATH=src/payload.config.ts payload generate:types && yarn copytypes:next",
    "copytypes:next": "copyfiles -u 1 src/payload-types.ts ../next-app/src/types/"
  },
  "dependencies": {
    "@azure/storage-blob": "^12.14.0",
    "@payloadcms/plugin-cloud-storage": "^1.0.16",
    "dotenv": "^16.1.4",
    "express": "^4.17.1",
    "payload": "^1.8.5",
    "payload-plugin-lexical": "^0.6.6",
    "react-colorful": "^5.6.1"
  },
  "devDependencies": {
    "@types/express": "^4.17.9",
    "@types/react": "^18.2.0",
    "copyfiles": "^2.4.1",
    "cross-env": "^7.0.3",
    "nodemon": "^2.0.6",
    "ts-node": "^10.9.1",
    "typescript": "^5.1.3"
  }
}
```
The code where I try to use the plugin:
```
import { cloudStorage } from '@payloadcms/plugin-cloud-storage'
import { LexicalPlugin } from 'payload-plugin-lexical'
import { Plugin } from 'payload/config'
import { CreateConfigFunction } from './createBuildConfig'
import { azureBlobStorageAdapter } from '@payloadcms/plugin-cloud-storage/azure'

export const plugins: CreateConfigFunction<Plugin[]> = ({
  isDev,
  cloudStorage: { connectionString, containerName, allowContainerCreate, baseURL },
}) => {
  const productionPlugins = isDev
    ? []
    : [
        cloudStorage({
          collections: {
            media: {
              adapter: azureBlobStorageAdapter({
                connectionString,
                containerName,
                allowContainerCreate,
                baseURL,
              }),
            },
          },
        }),
      ]

  return [LexicalPlugin({}), ...productionPlugins]
}
```
My node version: `18.15.0`
Please help to resolve these issues.

--- a/README.md
+++ b/README.md
@@ -37,6 +37,23 @@ export default buildConfig({
 });
 ```
 
+### Conditionally Enabling/Disabling
+
+The proper way to conditionally enable/disable this plugin is to use the `enabled` property.
+
+```ts
+cloudStorage({
+  enabled: process.env.MY_CONDITION === 'true',
+  collections: {
+    'my-collection-slug': {
+      adapter: theAdapterToUse, // see docs for the adapter you want to use
+    },
+  },
+}),
+```
+
+If the code is included *in any way in your config* but conditionally disabled in another fashion, you may run into issues such as `Webpack Build Error: Can't Resolve 'fs' and 'stream'` or similar because the plugin must be run at all times in order to properly extend the webpack config.
+
 ## Features
 
 **Adapter-based Implementation**

--- a/dev/src/payload.config.ts
+++ b/dev/src/payload.config.ts
@@ -83,7 +83,7 @@ export default buildConfig({
     outputFile: path.resolve(__dirname, 'payload-types.ts'),
   },
   plugins: [
-    // @ts-expect-error
+    // @ts-expect-error Conflicting types for relative package
     cloudStorage({
       collections: {
         media: {

--- a/src/plugin.ts
+++ b/src/plugin.ts
@@ -16,19 +16,26 @@ import { getFields } from './fields/getFields'
 
 export const cloudStorage =
   (pluginOptions: PluginOptions) =>
-  (config: Config): Config => {
-    const { collections: allCollectionOptions } = pluginOptions
+  (incomingConfig: Config): Config => {
+    const { collections: allCollectionOptions, enabled } = pluginOptions
+    const config = { ...incomingConfig }
 
-    const webpack = extendWebpackConfig({ options: pluginOptions, config })
+    const webpack = extendWebpackConfig({ options: pluginOptions, config: incomingConfig })
+
+    config.admin = {
+      ...(config.admin || {}),
+      webpack,
+    }
+
+    // Return early if disabled. Only webpack config mods are applied.
+    if (enabled === false) {
+      return config
+    }
 
     const initFunctions: Array<() => void> = []
 
     return {
       ...config,
-      admin: {
-        ...(config.admin || {}),
-        webpack,
-      },
       collections: (config.collections || []).map(existingCollection => {
         const options = allCollectionOptions[existingCollection.slug]
 

--- a/src/types.ts
+++ b/src/types.ts
@@ -69,5 +69,11 @@ export interface CollectionOptions {
 }
 
 export interface PluginOptions {
+  /**
+   * Whether or not to enable the plugin
+   *
+   * Default: true
+   */
+  enabled?: boolean
   collections: Record<string, CollectionOptions>
 }

